metric_metadata/cached/cached.go

```diff
@@ -15,68 +15,114 @@
+	clock             util.Clock         // Here so we can mock out in tests
 
+	// Cached items
+	getAllTagsCache      map[api.MetricKey]*CachedTagSetList // The cache of metric -> tags
+	getAllTagsCacheMutex sync.Mutex                          // Mutex for getAllTagsCache
+
+	// Cache Config
+	freshness  time.Duration // How long until cache entries become stale
+	timeToLive time.Duration // How long until cache entries become expired
+
+	// Queue
 	backgroundQueue chan func(metadata.Context) error // A channel that holds background requests.
+	queueMutex      sync.Mutex                        // Synchronizing mutex for the queue
 }
 
 // Config stores data needed to instantiate a CachedMetricMetadataAPI.
 type Config struct {
-	TimeToLive   time.Duration
+	Freshness    time.Duration
 	RequestLimit int
+	TimeToLive   time.Duration
+}
+
+// CachedTagSetList is an item in the cache.
+type CachedTagSetList struct {
+	TagSets []api.TagSet // The tagsets for this metric
+	Expiry  time.Time    // The time at which the cache entry expires
+	Stale   time.Time    // The time at which the cache entry becomes stale
+
+	sync.Mutex // Synchronizing mutex
+
+	inflight bool           // Indicates a request is already in flight
+	enqueued bool           // Indicates a request has been enqueued
+	wg       sync.WaitGroup // Synchronizing wait group
+
+	fetchError error // Fetch error from the last attempt
 }
 
 // NewMetricMetadataAPI creates a cached API given configuration and an underlying API object.
 func NewMetricMetadataAPI(apiInstance metadata.MetricAPI, config Config) *CachedMetricMetadataAPI {
 	requests := make(chan func(metadata.Context) error, config.RequestLimit)
+	if config.Freshness == 0 {
+		config.Freshness = config.TimeToLive
+	}
 	return &CachedMetricMetadataAPI{
 		metricMetadataAPI: apiInstance,
-		getAllTagsCache:   map[api.MetricKey]CachedTagSetList{},
+		clock:             util.RealClock{},
+		getAllTagsCache:   map[api.MetricKey]*CachedTagSetList{},
+		freshness:         config.Freshness,
 		timeToLive:        config.TimeToLive,
 		backgroundQueue:   requests,
 	}
 }
 
-// CachedTagSetList is an item in the cache.
-type CachedTagSetList struct {
-	TagSets []api.TagSet // The tagsets for this metric
-	Expiry  time.Time    // The time at which the cache entry expires
-}
-
-// Expired tells whether the entry has zero-time (meaning absent) or is out-of-date.
-func (c CachedTagSetList) Expired() bool {
-	return c.Expiry.IsZero() || c.Expiry.Before(time.Now())
-}
+// addBackgroundGetAllTagsRequest adds a job to update the lag list for the given
+// metric. Requires the client hold the lock for the item in the cache
+func (c *CachedMetricMetadataAPI) addBackgroundGetAllTagsRequest(metricKey api.MetricKey) {
+	c.queueMutex.Lock()
+	defer c.queueMutex.Unlock()
 
-func (c *CachedMetricMetadataAPI) addBackgroundGetAllTagsRequest(metric api.MetricKey) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
 	if len(c.backgroundQueue) < cap(c.backgroundQueue) {
```

Comment:
I realize now that it's more idiomatic for this to say 
```
if len(c.backgroundQueue) >= cap(c.backgroundQueue) {
     log.Warningf("Unable to enqueue a background GetAllTags lookup for %s due to a full queue", metricKey)
     return
}
```

to avoid the additional nesting